specviz/io/wizard/wizard.py

```diff
@@ -0,0 +1,483 @@
+                                          basedir=os.path.expanduser('~/.specviz'))[0]
+
+        if filename == '':
+            return
+
+        string = self.as_yaml()
+        with open(filename, 'w') as f:
+            f.write(string)
+
+
+class FITSImportWizard(BaseImportWizard):
+    dataset_label = 'HDU'
+
+    def as_yaml_dict(self, name=None):
+        """
+        Convert the current configuration to a dictionary that can then be
+        serialized to YAML
+        """
+
+        yaml_dict = OrderedDict()
+        yaml_dict['name'] = name or self.ui.loader_name.text()
+        yaml_dict['extension'] = ['fits']
+        if self.helper_disp.component_name.startswith('WCS::'):
+            yaml_dict['wcs'] = {
+                'hdu': self.helper_disp.dataset_name,
+            }
+        else:
+            yaml_dict['dispersion'] = {
+                'hdu': self.helper_disp.dataset_name,
+                'col': self.helper_disp.component_name,
+                'unit': self.helper_disp.unit
+            }
+            yaml_dict['wcs'] = {
+                'hdu': self.helper_disp.dataset_name
+            }
+        yaml_dict['data'] = {
+            'hdu': self.helper_data.dataset_name,
+            'col': self.helper_data.component_name,
+            'unit': self.helper_data.unit
+        }
+        yaml_dict['uncertainty'] = {
+            'hdu': self.helper_unce.dataset_name,
+            'col': self.helper_unce.component_name,
+            'type': self.ui.combo_uncertainty_type.currentData()
+        }
+        yaml_dict['meta'] = {'author': 'Wizard'}
+
+        return yaml_dict
+
+
+def open_wizard():
+
+    # NOTE: for now we only deal with FITS files, but this is where we would need
+    # to add different filters.
+
+    filters = ["FITS (*.fits)"]
+    filename, file_filter = compat.getopenfilename(filters=";;".join(filters))
+
+    if filename == '':
+        return
+
+    if file_filter == 'FITS (*.fits)':
+        dialog = FITSImportWizard(simplify_arrays(parse_fits(filename)))
+        val = dialog.exec_()
+    else:
+        raise NotImplementedError(file_filter)
+
+    if val == 0:
+        return
+
+    # Make temporary YAML file
+    yaml_file = tempfile.mktemp()
+    with open(yaml_file, 'w') as f:
+        f.write(dialog.as_yaml(name=str(uuid.uuid4())))
+
+    # Temporarily load YAML file
+    yaml_filter = load_yaml_reader(yaml_file)
+
+    # Emit signal to indicate that file should be read
+    dispatch.on_file_read.emit(file_name=filename,
+                               file_filter=yaml_filter)
+
+    # Remove YAML loader - this is a big hack at the moment since there is no
+    # official way in Astropy to unregister a reader/writer/identifier
+    # FIXME: how do we determine when the file has been read?
+    import time
+    time.sleep(1)
+    io_registry._readers.pop((yaml_filter, Spectrum1DRef))
+    io_registry._identifiers.pop((yaml_filter, Spectrum1DRef))
```

Comment:
There a couple ways off the top of my head. One would be to listen for `on_added_data` events, since those are called after a data set has been loaded from a file:

```python
dispatch.on_file_read.emit(file_name=filename, file_filter=yaml_filter)

def remove_yaml_filter(data):
    if data.name != filename:
        return

    io_registry._readers.pop((yaml_filter, Spectrum1DRef))
    io_registry._identifiers.pop((yaml_filter, Spectrum1DRef))

    dipatch.unregister_listener("on_added_data", remove_yaml_filter)

dispatch.register_listener("on_added_data",  remove_yaml_filter)
```

Alternatively, you could just spin up your own thread:

```python
from specviz.core.threads import FileLoadThread

file_load_thread = FileLoadThread()

# Connect to the status bar in the gui, if you like
file_load_thread.status.connect(
    dispatch.on_status_message.emit)

# Define some result handling function
def file_load_result(data):
    dispatch.on_add_data.emit(data=d)

    io_registry._readers.pop((yaml_filter, Spectrum1DRef))
    io_registry._identifiers.pop((yaml_filter, Spectrum1DRef))

# Hook up thread
file_load_thread.result.connect(file_load_result)

# Call and start thread
file_load_thread(filename, yaml_filter)
file_load_thread.start()
```